File: app/src/main/java/org/neidhardt/dynamicsoundboard/soundmanagement_old/views/RenameSoundFileDialog.java
package org.neidhardt.dynamicsoundboard.soundmanagement_old.views;

import android.annotation.SuppressLint;
import android.app.AlertDialog;
import android.app.Dialog;
import android.app.FragmentManager;
import android.net.Uri;
import android.os.Bundle;
import android.view.View;
import android.widget.CheckBox;
import android.widget.Toast;
import de.greenrobot.event.EventBus;
import org.neidhardt.dynamicsoundboard.R;
import org.neidhardt.dynamicsoundboard.dao.MediaPlayerData;
import org.neidhardt.dynamicsoundboard.mediaplayer.EnhancedMediaPlayer;
import org.neidhardt.dynamicsoundboard.misc.FileUtils;
import org.neidhardt.dynamicsoundboard.misc.Logger;
import org.neidhardt.dynamicsoundboard.navigationdrawer.playlist.views.Playlist;
import org.neidhardt.dynamicsoundboard.soundmanagement_old.events.PlaylistChangedEvent;
import org.neidhardt.dynamicsoundboard.soundmanagement_old.model.SoundDataModel;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Set;

/**
 * File created by eric.neidhardt on 12.04.2015.
 */
public class RenameSoundFileDialog extends SoundSettingsBaseDialog implements View.OnClickListener
{
	public static final String TAG = RenameSoundFileDialog.class.getName();

	private CheckBox renameAllOccurrences;
	private MediaPlayerData playerData;
	private List<EnhancedMediaPlayer> playersWithMatchingUri;

	public static void showInstance(FragmentManager manager, MediaPlayerData playerData)
	{
		RenameSoundFileDialog dialog = new RenameSoundFileDialog();
		addArguments(dialog, playerData.getPlayerId(), playerData.getFragmentTag());
		dialog.show(manager, TAG);
	}

	@Override
	public Dialog onCreateDialog(Bundle savedInstanceState)
	{
		@SuppressLint("InflateParams")
		View view = this.getActivity().getLayoutInflater().inflate(R.layout.dialog_rename_sound_file_layout, null);
		this.renameAllOccurrences = (CheckBox) view.findViewById(R.id.cb_rename_all_occurrences);

		view.findViewById(R.id.b_ok).setOnClickListener(this);
		view.findViewById(R.id.b_cancel).setOnClickListener(this);

		AlertDialog.Builder builder = new AlertDialog.Builder(this.getActivity());
		builder.setView(view);

		if (this.player != null)
			this.setMediaPlayerData(this.player.getMediaPlayerData());

		return builder.create();
	}

	void setMediaPlayerData(MediaPlayerData playerData)
	{
		this.playerData = playerData;

		this.playersWithMatchingUri = this.getPlayersWithMatchingUri(this.playerData.getUri());
		if (playersWithMatchingUri.size() > 1)
		{
			this.renameAllOccurrences.setVisibility(View.VISIBLE);
			this.renameAllOccurrences.setText(this.renameAllOccurrences.getText().toString().replace("{%s0}", Integer.toString(playersWithMatchingUri.size())));
		}
		else
			this.renameAllOccurrences.setVisibility(View.GONE);
	}

	@Override
	public void onClick(View v)
	{
		switch (v.getId())
		{
			case R.id.b_cancel:
				this.dismiss();
				break;
			case R.id.b_ok:
				this.deliverResult(Uri.parse(this.playerData.getUri()), this.playerData.getLabel(), this.renameAllOccurrences.isChecked());
				this.dismiss();
				break;
		}
	}

	void deliverResult(Uri fileUriToRename, String newFileLabel, boolean renameAllOccurrences) {
		File fileToRename = FileUtils.getFileForUri(this.getActivity(), fileUriToRename);
		if (fileToRename == null) {
			this.showErrorRenameFile();
			return;
		}

		String newFilePath = fileToRename.getAbsolutePath().replace(fileToRename.getName(), "") + this.appendFileTypeToNewPath(newFileLabel, fileToRename.getName());
		if (newFilePath.equals(fileToRename.getAbsolutePath())) {
			Logger.d(TAG, "old name and new name are equal, nothing to be done");
			return;
		}

		File newFile = new File(newFilePath);
		boolean success = fileToRename.renameTo(newFile);
		if (!success) {
			this.showErrorRenameFile();
			return;
		}

		String newUri = Uri.fromFile(newFile).toString();
		for (EnhancedMediaPlayer player : this.playersWithMatchingUri) {
			if (!this.setUriForPlayer(player, newUri))
				this.showErrorRenameFile();

			if (renameAllOccurrences) {
				player.getMediaPlayerData().setLabel(newFileLabel);
				player.getMediaPlayerData().setItemWasAltered();
			}

			if (player.getMediaPlayerData().getFragmentTag().equals(Playlist.TAG))
				EventBus.getDefault().post(new PlaylistChangedEvent());
			else
				this.notifyFragment(player.getMediaPlayerData().getFragmentTag());
		}
	}

	private void showErrorRenameFile()
	{
		if (this.getActivity() != null)
			Toast.makeText(this.getActivity(), R.string.dialog_rename_sound_toast_player_not_updated, Toast.LENGTH_SHORT).show();
	}

	String appendFileTypeToNewPath(String newNameFilePath, String oldFilePath)
	{
		if (newNameFilePath == null || oldFilePath == null)
			throw new NullPointerException(TAG + ": cannot create new file name, either old name or new name is null");

		String[] segments = oldFilePath.split("\\.");
		if (segments.length > 1) {
			String fileType = segments[segments.length - 1];
			return newNameFilePath + "." + fileType;
		}

		return newNameFilePath;
	}

	boolean setUriForPlayer(EnhancedMediaPlayer player, String uri)
	{
		SoundDataModel model = this.getServiceManagerFragment();
		try
		{
			player.setSoundUri(uri);
			return true;
		}
		catch (IOException e)
		{
			Logger.e(TAG, e.getMessage());
			if (player.getMediaPlayerData().getFragmentTag().equals(Playlist.TAG))
				model.removeSoundsFromPlaylist(Collections.singletonList(player));
			else
				model.removeSounds(Collections.singletonList(player));
			return false;
		}
	}

	List<EnhancedMediaPlayer> getPlayersWithMatchingUri(String uri)
	{
		SoundDataModel model = this.getServiceManagerFragment();
		List<EnhancedMediaPlayer> players = new ArrayList<>();

		for (EnhancedMediaPlayer player : model.getPlayList())
		{
			if (player.getMediaPlayerData().getUri().equals(uri))
				players.add(player);
		}

		Set<String> fragments = model.getSounds().keySet();
		for (String fragment : fragments)
		{
			List<EnhancedMediaPlayer> soundsInFragment = model.getSoundsInFragment(fragment);
			for (EnhancedMediaPlayer player : soundsInFragment)
			{
				if (player.getMediaPlayerData().getUri().equals(uri))
					players.add(player);
			}
		}

		return players;
	}
}

File: app/src/main/kotlin/org/neidhardt/dynamicsoundboard/soundmanagement/dialog/RenameSoundFileDialog.kt
package org.neidhardt.dynamicsoundboard.soundmanagement.dialog

import android.annotation.SuppressLint
import android.app.Dialog
import android.net.Uri
import android.os.Bundle
import android.support.v4.app.FragmentManager
import android.support.v7.app.AlertDialog
import android.view.View
import android.widget.CheckBox
import android.widget.TextView
import android.widget.Toast
import org.greenrobot.eventbus.EventBus
import org.neidhardt.dynamicsoundboard.R
import org.neidhardt.dynamicsoundboard.dao.MediaPlayerData
import org.neidhardt.dynamicsoundboard.mediaplayer.MediaPlayerController
import org.neidhardt.dynamicsoundboard.mediaplayer.PlaylistTAG
import org.neidhardt.dynamicsoundboard.misc.Logger
import org.neidhardt.dynamicsoundboard.misc.getFileForUri
import org.neidhardt.dynamicsoundboard.soundmanagement.events.PlaylistChangedEvent
import org.neidhardt.dynamicsoundboard.soundmanagement.events.SoundChangedEvent
import org.neidhardt.dynamicsoundboard.soundmanagement.model.SoundsDataAccess
import org.neidhardt.dynamicsoundboard.soundmanagement.model.SoundsDataStorage
import org.neidhardt.dynamicsoundboard.soundmanagement.views.SoundSettingsBaseDialog
import java.io.File
import java.io.IOException
import java.util.*

/**
 * File created by eric.neidhardt on 06.07.2015.
 */
class RenameSoundFileDialog : SoundSettingsBaseDialog
{
	override var player: MediaPlayerController? = null
	override var fragmentTag: String? = null

	private val TAG = javaClass.name

	constructor() : super()

	constructor(manager: FragmentManager, playerData: MediaPlayerData) : super()
	{
		SoundSettingsBaseDialog.addArguments(this, playerData.playerId, playerData.fragmentTag)
		this.show(manager, TAG)
	}

	override fun onCreateDialog(savedInstanceState: Bundle?): Dialog
	{
		@SuppressLint("InflateParams")
		val view = this.activity.layoutInflater.inflate(R.layout.dialog_rename_sound_file_layout, null)

		val presenter = RenameSoundFileDialogPresenter(
				playerData = this.player!!.mediaPlayerData,
				soundsDataAccess = this.soundsDataAccess,
				soundsDataStorage = this.soundsDataStorage,
				dialog = this,
				newName = view.findViewById(R.id.tv_new_name) as TextView,
				renameAllOccurrences = view.findViewById(R.id.cb_rename_all_occurrences) as CheckBox
		)

		return AlertDialog.Builder(context).apply {
			this.setView(view)
			this.setPositiveButton(R.string.dialog_rename, { dialogInterface, i ->
				presenter.rename()
			})
			this.setNegativeButton(R.string.dialog_cancel, { dialogInterface, i -> dismiss() })
		}.create()
	}
}

private class RenameSoundFileDialogPresenter
(
		private val playerData: MediaPlayerData,
		private val soundsDataAccess: SoundsDataAccess,
		private val soundsDataStorage: SoundsDataStorage,
		private val dialog: RenameSoundFileDialog,
		private val newName : TextView,
		private val renameAllOccurrences: CheckBox
)
{
	private val TAG = javaClass.name

	private var playersWithMatchingUri: List<MediaPlayerController>? = null

	init
	{
		this.playersWithMatchingUri = this.getPlayersWithMatchingUri(this.playerData.uri)
		if (playersWithMatchingUri!!.size > 1)
		{
			this.renameAllOccurrences.visibility = View.VISIBLE
			this.renameAllOccurrences.text = this.renameAllOccurrences.text.toString()
					.replace("{%s0}", Integer.toString(playersWithMatchingUri!!.size))
		}
		else
			this.renameAllOccurrences.visibility = View.GONE

		val currentFile = Uri.parse(this.playerData.uri).getFileForUri()
		val currentFileName = currentFile?.name

		this.newName.text = this.appendFileTypeToNewPath(this.playerData.label, currentFileName)
	}

	private fun getPlayersWithMatchingUri(uri: String): List<MediaPlayerController>
	{
		val players = ArrayList<MediaPlayerController>()

		val playlist = this.soundsDataAccess.playlist
		for (player in playlist)
		{
			if (player.mediaPlayerData.uri == uri)
				players.add(player)
		}

		val fragments = this.soundsDataAccess.sounds.keys
		for (fragment in fragments)
		{
			val soundsInFragment = this.soundsDataAccess.getSoundsInFragment(fragment)
			for (player in soundsInFragment)
			{
				if (player.mediaPlayerData.uri == uri)
					players.add(player)
			}
		}

		return players
	}

	internal fun rename()
	{
		val uri = Uri.parse(this.playerData.uri)
		val label = this.playerData.label
		val renameAllOccurrences = this.renameAllOccurrences.isChecked

		this.deliverResult(uri, label, renameAllOccurrences)

		this.dialog.dismiss()
	}

	private fun deliverResult(fileUriToRename: Uri, newFileLabel: String, renameAllOccurrences: Boolean)
	{
		val fileToRename = fileUriToRename.getFileForUri()
		if (fileToRename == null)
		{
			this.showErrorRenameFile()
			return
		}

		val newFilePath = fileToRename.absolutePath.replace(fileToRename.name, "") + this.appendFileTypeToNewPath(newFileLabel, fileToRename.name)
		if (newFilePath == fileToRename.absolutePath)
		{
			Logger.d(TAG, "old name and new name are equal, nothing to be done")
			return
		}

		val newFile = File(newFilePath)
		val success = fileToRename.renameTo(newFile)
		if (!success)
		{
			this.showErrorRenameFile()
			return
		}

		val newUri = Uri.fromFile(newFile).toString()
		for (player in this.playersWithMatchingUri!!)
		{
			if (!this.setUriForPlayer(player, newUri))
				this.showErrorRenameFile()

			if (renameAllOccurrences)
			{
				player.mediaPlayerData.label = newFileLabel
				player.mediaPlayerData.updateItemInDatabaseAsync()
			}

			if (player.mediaPlayerData.fragmentTag == PlaylistTAG)
				EventBus.getDefault().post(PlaylistChangedEvent())
			else
				EventBus.getDefault().post(SoundChangedEvent(player))
		}
	}

	private fun showErrorRenameFile()
	{
		if (this.dialog.activity != null)
			Toast.makeText(this.dialog.activity, R.string.dialog_rename_sound_toast_player_not_updated, Toast.LENGTH_SHORT).show()
	}

	private fun appendFileTypeToNewPath(newNameFilePath: String?, oldFilePath: String?): String
	{
		if (newNameFilePath == null || oldFilePath == null)
			throw NullPointerException(TAG + ": cannot create new file name, either old name or new name is null")

		val segments = oldFilePath.split("\\.".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
		if (segments.size > 1) {
			val fileType = segments[segments.size - 1]
			return newNameFilePath + "." + fileType
		}

		return newNameFilePath
	}

	private fun setUriForPlayer(player: MediaPlayerController, uri: String): Boolean
	{
		try
		{
			player.setSoundUri(uri)
			return true
		}
		catch (e: IOException)
		{
			Logger.e(TAG, e.message)
			if (player.mediaPlayerData.fragmentTag == PlaylistTAG)
				this.soundsDataStorage.removeSoundsFromPlaylist(listOf(player))
			else
				this.soundsDataStorage.removeSounds(listOf(player))
			return false
		}
	}
}
